src/Marten/Internal/ProviderGraph.cs

```diff
@@ -23,7 +24,10 @@ public ProviderGraph(StoreOptions options)
 
     public void Append<T>(DocumentProvider<T> provider) where T : notnull
     {
-        _storage = _storage.AddOrUpdate(typeof(T), provider);
+        lock (_storageLock)
```

Comment:
Don't do the lock here, it's unnecessary w/ the immutable hash map

Reply:
I'm not sure how `ImHashMap`s work exactly, but the reason I also locked it is because I thought that if two threads enter this method concurrently, then they call `AddOrUpdate` on the same instance. Then they each would create a new map based on that same map instance, resulting in two separate instances that do not contain the item the other thread tried to add. The assignment is atomic, so whoever assigns to `_storage` last "wins".

Reply:
Maybe I'll just add that as another test case and see if that can happen. Although tests of that nature aren't the most reliable.

Reply:
Seems that it is at least possible:
![image](https://github.com/JasperFx/marten/assets/24227002/fda2bcfc-3919-41d6-91be-298868c663f9)
I can't speak to how likely that is in practice.